File: src/main/java/marioandweegee3/ironbarrels2/IronBarrels.java
package marioandweegee3.ironbarrels2;

import marioandweegee3.ironbarrels2.block.BigBarrelBlock;
import marioandweegee3.ironbarrels2.block.entity.BarrelEntities;
import marioandweegee3.ironbarrels2.block.entity.BigBarrelEntity;
import marioandweegee3.ironbarrels2.item.UpgradeKit;
import marioandweegee3.ironbarrels2.item.VanillaBarrelUpgradeKit;
import marioandweegee3.ml3api.registry.RegistryHelper;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.block.Block;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import org.apache.logging.log4j.Logger;

import java.util.HashMap;
import java.util.Map;

public class IronBarrels implements ModInitializer {
    public static final String modid = "ironbarrels2";
    public static final RegistryHelper helper = new RegistryHelper(modid);
    public static final Logger logger = helper.logger;

    public static final BlockEntityType<BigBarrelEntity> IRON_BARREL = BlockEntityType.Builder
            .create(() -> BarrelEntities.barrel(0), BigBarrelBlock.BARRELS[0]).build(null);
    public static final BlockEntityType<BigBarrelEntity> GOLD_BARREL = BlockEntityType.Builder
            .create(() -> BarrelEntities.barrel(1), BigBarrelBlock.BARRELS[1]).build(null);
    public static final BlockEntityType<BigBarrelEntity> DIAMOND_BARREL = BlockEntityType.Builder
            .create(() -> BarrelEntities.barrel(2), BigBarrelBlock.BARRELS[2]).build(null);
    public static final BlockEntityType<BigBarrelEntity> OBSIDIAN_BARREL = BlockEntityType.Builder
            .create(() -> BarrelEntities.barrel(3), BigBarrelBlock.BARRELS[3]).build(null);
    public static final BlockEntityType<BigBarrelEntity> COPPER_BARREL = BlockEntityType.Builder
            .create(() -> BarrelEntities.barrel(4), BigBarrelBlock.BARRELS[4]).build(null);
    public static final BlockEntityType<BigBarrelEntity> SILVER_BARREL = BlockEntityType.Builder
            .create(() -> BarrelEntities.barrel(5), BigBarrelBlock.BARRELS[5]).build(null);

    public static final Item debug_item = new Item(new Item.Settings().maxCount(1));

    @Override
    public void onInitialize() {

        helper.registerBlockEntity("iron_barrel", IRON_BARREL);
        helper.registerBlockEntity("gold_barrel", GOLD_BARREL);
        helper.registerBlockEntity("diamond_barrel", DIAMOND_BARREL);
        helper.registerBlockEntity("obsidian_barrel", OBSIDIAN_BARREL);
        helper.registerBlockEntity("copper_barrel", COPPER_BARREL);
        helper.registerBlockEntity("silver_barrel", SILVER_BARREL);

        Map<String, Block> barrels = new HashMap<>();
        barrels.put("iron_barrel", BigBarrelBlock.BARRELS[0]);
        barrels.put("gold_barrel", BigBarrelBlock.BARRELS[1]);
        barrels.put("diamond_barrel", BigBarrelBlock.BARRELS[2]);
        barrels.put("obsidian_barrel", BigBarrelBlock.BARRELS[3]);
        barrels.put("copper_barrel", BigBarrelBlock.BARRELS[4]);
        barrels.put("silver_barrel", BigBarrelBlock.BARRELS[5]);
        helper.registerAllBlocks(barrels, ItemGroup.DECORATIONS);

        helper.registerItem("wood_copper_kit", new VanillaBarrelUpgradeKit(BigBarrelBlock.BARRELS[4]));
        helper.registerItem("wood_iron_kit", new VanillaBarrelUpgradeKit(BigBarrelBlock.BARRELS[0]));
        helper.registerItem("wood_silver_kit", new VanillaBarrelUpgradeKit(BigBarrelBlock.BARRELS[5]));
        helper.registerItem("wood_gold_kit", new VanillaBarrelUpgradeKit(BigBarrelBlock.BARRELS[1]));
        helper.registerItem("wood_diamond_kit", new VanillaBarrelUpgradeKit(BigBarrelBlock.BARRELS[2]));
        helper.registerItem("wood_obsidian_kit", new VanillaBarrelUpgradeKit(BigBarrelBlock.BARRELS[3]));
        helper.registerItem("copper_iron_kit", new UpgradeKit(BigBarrelBlock.BARRELS[4], BigBarrelBlock.BARRELS[0]));
        helper.registerItem("copper_silver_kit", new UpgradeKit(BigBarrelBlock.BARRELS[4], BigBarrelBlock.BARRELS[5]));
        helper.registerItem("copper_gold_kit", new UpgradeKit(BigBarrelBlock.BARRELS[4], BigBarrelBlock.BARRELS[1]));
        helper.registerItem("copper_diamond_kit", new UpgradeKit(BigBarrelBlock.BARRELS[4], BigBarrelBlock.BARRELS[2]));
        helper.registerItem("copper_obsidian_kit",
                new UpgradeKit(BigBarrelBlock.BARRELS[4], BigBarrelBlock.BARRELS[3]));
        helper.registerItem("iron_silver_kit", new UpgradeKit(BigBarrelBlock.BARRELS[0], BigBarrelBlock.BARRELS[5]));
        helper.registerItem("iron_gold_kit", new UpgradeKit(BigBarrelBlock.BARRELS[0], BigBarrelBlock.BARRELS[1]));
        helper.registerItem("iron_diamond_kit", new UpgradeKit(BigBarrelBlock.BARRELS[0], BigBarrelBlock.BARRELS[2]));
        helper.registerItem("iron_obsidian_kit", new UpgradeKit(BigBarrelBlock.BARRELS[0], BigBarrelBlock.BARRELS[3]));
        helper.registerItem("silver_gold_kit", new UpgradeKit(BigBarrelBlock.BARRELS[5], BigBarrelBlock.BARRELS[1]));
        helper.registerItem("silver_diamond_kit", new UpgradeKit(BigBarrelBlock.BARRELS[5], BigBarrelBlock.BARRELS[2]));
        helper.registerItem("silver_obsidian_kit",
                new UpgradeKit(BigBarrelBlock.BARRELS[5], BigBarrelBlock.BARRELS[3]));
        helper.registerItem("gold_diamond_kit", new UpgradeKit(BigBarrelBlock.BARRELS[1], BigBarrelBlock.BARRELS[2]));
        helper.registerItem("gold_obsidian_kit", new UpgradeKit(BigBarrelBlock.BARRELS[1], BigBarrelBlock.BARRELS[3]));
        helper.registerItem("diamond_obsidian_kit",
                new UpgradeKit(BigBarrelBlock.BARRELS[2], BigBarrelBlock.BARRELS[3]));

        if (FabricLoader.getInstance().isDevelopmentEnvironment()) {
            helper.registerItem("debug_item", debug_item);
        }
    }

    public static void log(String message) {
        logger.info("[" + modid + "] " + message);
    }

}

File: src/main/scala/marioandweegee3/ironbarrels3/IronBarrels.scala
package marioandweegee3.ironbarrels3

import com.swordglowsblue.artifice.api.Artifice
import com.swordglowsblue.artifice.api.ArtificeResourcePack.ServerResourcePackBuilder
import com.swordglowsblue.artifice.api.builder.data.TagBuilder
import com.swordglowsblue.artifice.api.builder.data.recipe.ShapedRecipeBuilder
import marioandweegee3.ironbarrels3.block.BigBarrelBlock
import marioandweegee3.ironbarrels3.block.entity.BigBarrelEntity
import marioandweegee3.ironbarrels3.items.{UpgradeKit, VanillaUpgradeKit}
import marioandweegee3.ml3api.registry.RegistryHelper
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.block.FabricBlockSettings
import net.fabricmc.fabric.api.tools.FabricToolTags
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.block.{Block, Blocks}
import net.minecraft.item.{Item, ItemGroup}
import net.minecraft.tag.ItemTags
import net.minecraft.util.Identifier

import scala.collection.mutable
import scala.jdk.CollectionConverters._

object IronBarrels extends ModInitializer {
	val barrelSettings:Block.Settings = FabricBlockSettings.copy(Blocks.IRON_DOOR).breakByTool(FabricToolTags.PICKAXES).breakByHand(true).build()
	
	val blocks: Map[String, Block] = Map[String, Block](
		("copper_barrel", new BigBarrelBlock(barrelSettings, 0)),
		("iron_barrel", new BigBarrelBlock(barrelSettings, 1)),
		("silver_barrel", new BigBarrelBlock(barrelSettings, 2)),
		("gold_barrel", new BigBarrelBlock(barrelSettings, 3)),
		("diamond_barrel", new BigBarrelBlock(barrelSettings, 4))
	)
	
	val COPPER:BlockEntityType[BigBarrelEntity] = new BlockEntityType[BigBarrelEntity](
		() => new BigBarrelEntity(BarrelType.COPPER),
		mutable.HashSet[Block](
			blocks("copper_barrel")
		).asJava,
		null
	)
	val IRON:BlockEntityType[BigBarrelEntity] = new BlockEntityType[BigBarrelEntity](
		() => new BigBarrelEntity(BarrelType.IRON),
		mutable.HashSet[Block](
			blocks("iron_barrel")
		).asJava,
		null
	)
	val SILVER:BlockEntityType[BigBarrelEntity] = new BlockEntityType[BigBarrelEntity](
		() => new BigBarrelEntity(BarrelType.SILVER),
		mutable.HashSet[Block](
			blocks("silver_barrel")
		).asJava,
		null
	)
	val GOLD:BlockEntityType[BigBarrelEntity] = new BlockEntityType[BigBarrelEntity](
		() => new BigBarrelEntity(BarrelType.GOLD),
		mutable.HashSet[Block](
			blocks("gold_barrel")
		).asJava,
		null
	)
	val DIAMOND:BlockEntityType[BigBarrelEntity] = new BlockEntityType[BigBarrelEntity](
		() => new BigBarrelEntity(BarrelType.DIAMOND),
		mutable.HashSet[Block](
			blocks("diamond_barrel")
		).asJava,
		null
	)
	
	val helper = new RegistryHelper("ironbarrels3")
	
	override def onInitialize(): Unit = {
		helper.registerAllBlocks(blocks.asJava, ItemGroup.DECORATIONS)
		
		helper.registerBlockEntity("copper_barrel", COPPER)
		helper.registerBlockEntity("iron_barrel", IRON)
		helper.registerBlockEntity("silver_barrel", SILVER)
		helper.registerBlockEntity("gold_barrel", GOLD)
		helper.registerBlockEntity("diamond_barrel", DIAMOND)
		
		Artifice.registerData(helper.makeId("data"), (pack:ServerResourcePackBuilder) => {
			pack.addItemTag("c:barrels", (tag:TagBuilder) => {
				tag.value("barrel")
			})
			pack.addBlockTag("c:barrels", (tag:TagBuilder) => {
				tag.value("barrel")
			})
			if(itemTagExists("c:copper_ingot")){
				pack.addShapedRecipe("ironbarrels3:wood_to_copper_kit", (recipe:ShapedRecipeBuilder) => {
					recipe.pattern(
						"cpc"
					)
					recipe.ingredientTag('p', "planks")
					recipe.ingredientTag('c', "c:copper_ingot")
					recipe.result("ironbarrels3:wood_to_copper_kit", 1)
				})
				pack.addShapedRecipe("ironbarrels3:copper_barrel", (recipe:ShapedRecipeBuilder) => {
					recipe.pattern(
						"cpc"
					)
					recipe.ingredientTag('p', "c:barrels")
					recipe.ingredientTag('c', "c:copper_ingot")
					recipe.result("ironbarrels3:copper_barrel", 1)
				})
			}
		})
		
		val kitSettings = new Item.Settings().group(ItemGroup.MISC)
		
		for(b <- BarrelType.values()){
			for(b2 <- BarrelType.values().filter((barrel:BarrelType) => b.compareTo(barrel) < 0)){
				val type1 = b.name().toLowerCase
				val type2 = b2.name().toLowerCase
				
				helper.registerItem(s"${type1}_to_${type2}_kit", new UpgradeKit(kitSettings, b, b2))
			}
			helper.registerItem(s"wood_to_${b.name().toLowerCase}_kit", new VanillaUpgradeKit(kitSettings, b))
		}
	}
	
	def itemTagExists(id:Identifier):Boolean = ItemTags.getContainer.getKeys.contains(id)
	
	@inline
	implicit def stringsToId(strings:(String, String)):Identifier = new Identifier(strings._1, strings._2)
	
	@inline
	implicit def stringToId(string: String):Identifier = new Identifier(string)
	
	@inline
	implicit def idToStrings(identifier: Identifier):(String, String) = (identifier.getNamespace, identifier.getPath)
}
